### PR TITLE
LibWeb: Compute scrollable overflow via containing block instead of children

### PR DESCRIPTION
**Pull request by ebanner**
Instead of using child boxes to compute scrollable overflow for the box, we use descendants which have the box as their containing block.

Fixes https://github.com/LadybirdBrowser/ladybird/issues/718

## Failing Tests

- [x] abspos-flexbox-with-auto-width.html
- [x] misc/create-iframes-using-innerhtml.html
- [x] abspos-box-with-replaced-element.html
- [x] abspos-flexbox-with-auto-height.html
- [x] pseudo-element-with-custom-properties.html
- [x] automatic-width-of-non-replaced-abspos-element-must-not-be-negative.html
- [x] abspos-with-static-position-in-one-axis.html
- [x] abspos-image-with-min-height-constraint.html
- [x] css-logical-inset-properties.html
- [x] flex-abspos-item-with-preceding-whitespace.html
- [x] table-fixup-with-rowspan.html
- [x] table/basic.html
- [x] table/size.html
- [x] table/row-outer-size-with-computed-size.html
- [x] table/table-formation-with-rowspan-in-the-middle.html
- [x] table/propagate-style-update-to-wrapper.html
- [x] table/rowspan-with-trailing-characters.html
- [x] table/bottom-caption.html
- [x] table/align-top-and-bottom.html
- [x] table/nested-table-box-width.html
- [x] table/long-caption-increases-width.html
- [x] table/border-collapse-is-inherited.html
- [x] table/rowspan.html
- [x] table/avoid-div-by-zero-in-table-measures.html
  - #1790
- [x] table/row-span-and-nested-tables.html
- [x] table/border-spacing-rowspan.html
- [x] table/abspos-pseudo-element-inside-table.html
- [x] table/clip-spans-to-table-end.html
- [x] link-sheet.html
- [x] abspos-flex-container-with-auto-height.html
- [x] blockify-layout-internal-box-without-crashing.html
- [x] block-and-inline/max-width-percentage-100.html
- [x] block-and-inline/max-width-must-not-expand-element.html
- [x] block-and-inline/max-width-percentage-100-border-box.html
- [x] block-and-inline/width-auto-margins-set-zero-if-containing-size-smaller.html
- [x] block-and-inline/automatic-size-of-block-container-with-inline-children-and-nonempty-floating-child.html
- [x] block-and-inline/abspos-with-auto-margins-in-inline-axis.html
- [x] block-and-inline/abspos-grid-with-definite-width.html
- [x] automatic-height-of-non-replaced-abspos-element-must-not-be-negative.html
- [x] abspos-with-percentage-insets.html
- [x] flex/intrinsic-cross-size-contribution-with-main-size-constraint.html
- [x] flex/abspos-flex-child-with-percentage-height.html
- [x] flex/abspos-flex-child-static-position-with-padding-on-flex-container.html
- [x] flex/abspos-flex-child-static-position-with-align-items.html
- [x] abspos-box-with-block-level-sibling.html
- [x] input-image.html
- [x] grid/abspos-with-grid-container-as-parent.html

### CODE REVIEW

**Comment by ebanner** [comment]
This PR causes a whole bunch of tests to break. Looking at them, it's not clear to me that the tests are correct, though.

Most (all?) of the tests that break essentially look like this:

```
Test failed: file:///Users/edward/Code/ladybird/Tests/LibWeb/Layout/input/table/clip-spans-to-table-end.html
205/1304: Layout/input/table/table-cellpadd--- /Users/edward/Code/ladybird/Tests/LibWeb/Layout/expected/./table/table-cellpadding.txt
+++ /Users/edward/Code/ladybird/Tests/LibWeb/Layout/expected/./table/table-cellpadding.txt
@@ -25,8 +25,8 @@
     PaintableWithLines (BlockContainer<BODY>) [8,8 784x204]
       PaintableWithLines (TableWrapper(anonymous)) [8,8 498.1875x204]
         PaintableBox (Box<TABLE>) [8,8 498.1875x204]
-          PaintableBox (Box<TBODY>) [9,9 488.1875x198] overflow: [9,9 494.1875x200]
-            PaintableBox (Box<TR>) [11,11 488.1875x198] overflow: [11,11 492.1875x198]
+          PaintableBox (Box<TBODY>) [9,9 488.1875x198]
+            PaintableBox (Box<TR>) [11,11 488.1875x198]
               PaintableWithLines (BlockContainer<TD>) [11,11 146.640625x198]
                 TextPaintable (TextNode<#text>)
               PaintableWithLines (BlockContainer<TD>) [159.640625,11 165.4375x198]
```

That is, the overflow is removed.

I'm reasonably sure there should be no overflow though. In this test for instance, TBODY and TR and not containing blocks, so they can't have any overflow (is that right?). Also, Chrome doesn't compute any overflow for TBODY nor TR. You can verify this by adding:

```css
tbody, tr {
    overflow: scroll;
}
```

and verifying that Chrome does not display any scrollbars (whereas ladybird does).

I looked at a few more of the table tests and they follow this pattern and can be verified by setting `overflow: scroll`. There are other tests like `block-and-inline` where overflow is present on HTML and BODY, but in most cases HTML and BODY are not containing blocks. I also looked at some of the `flex` tests and a similar thing follows for the DIVs (where they are mostly not containing blocks but are getting overflow).

**Comment by kalenikaliaksandr** [comment]
> I'm reasonably sure there should be no overflow though. In this test for instance, TBODY and TR and not containing blocks, so they can't have any overflow (is that right?). Also, Chrome doesn't compute any overflow for TBODY nor TR. You can verify this by adding:

you are right, in this specific example you provided, removed overflow is improvement. so you would have to simply update a test expectation in that case. though I bet there might be some valid regressions, particularly when CSS transform is involved. the last time I attempted to make similar change, I discovered that if we don't also resolve this FIXME, it will cause some regression in the ref-tests.
`FIXME: accounting for transforms by projecting each box onto the plane of the element that establishes its 3D rendering context. [CSS3-TRANSFORMS]`

**Comment by ebanner** [comment]
Most (all?) of the table tests are failing because `<TBODY>` are getting overflow when they presumably shouldn't be: https://github.com/LadybirdBrowser/ladybird/issues/1525

**Comment by ebanner** [comment]
@kalenikaliaksandr I don't see any failures in the ref-tests. That means we don't have to address that FIXME for this issue..?

**Comment by kalenikaliaksandr** [comment]
> @kalenikaliaksandr I don't see any failures in the ref-tests. That means we don't have to address that FIXME for this issue..?

what FIXME do you mean exactly? btw please rebase your branch.

**Comment by ebanner** [comment]
I was responding to this part of your comment:

> though I bet there might be some valid regressions, particularly when CSS transform is involved. the last time I attempted to make similar change, I discovered that if we don't also resolve this FIXME, it will cause some regression in the ref-tests.
FIXME: accounting for transforms by projecting each box onto the plane of the element that establishes its 3D rendering context. [CSS3-TRANSFORMS]

**Comment by kalenikaliaksandr** [comment]
> I was responding to this part of your comment:
> 
> > though I bet there might be some valid regressions, particularly when CSS transform is involved. the last time I attempted to make similar change, I discovered that if we don't also resolve this FIXME, it will cause some regression in the ref-tests.
> > FIXME: accounting for transforms by projecting each box onto the plane of the element that establishes its 3D rendering context. [CSS3-TRANSFORMS]

yes, if this change does not break any tests then it's fine to leave the FIXME.

**Comment by kalenikaliaksandr** [comment]
> I'm not quite sure what you mean by rebase my branch. Rebase it off of..? I don't see anything that looks off.

I mean rebase onto master branch to include latest changes from there by doing `git rebase master`.

**Review by kalenikaliaksandr — APPROVED** [review]
I tested change locally and it seems good. likely it's going to result in overflow measuring being heavy in profiles because we will do full subtree traversal for each child, but that could be fixed.

**Comment by ebanner** [comment]
ok done 👍

(btw I'm curious what you saw on the PR that led you to ask me to rebase off master)

**Comment by kalenikaliaksandr** [comment]
> ok done 👍
> 
> (btw I'm curious what you saw on the PR that led you to ask me to rebase off master)

I seen InlinePaintable in text expectations which is gone by now, so I was a bit confused. apparently it does not introduce any conflicts so that's fine.

**Comment by ebanner** [comment]
> I tested change locally and it seems good. likely it's going to result in overflow measuring being heavy in profiles because we will do full subtree traversal for each child, but that could be fixed.

What are you thinking for a fix? The most obvious one that jumps out to me is to have a box point to all the elements for which it's the containing block (so it doesn't have to traverse the whole subtree every time to find them).

**Comment by kalenikaliaksandr** [comment]
> > I tested change locally and it seems good. likely it's going to result in overflow measuring being heavy in profiles because we will do full subtree traversal for each child, but that could be fixed.
> 
> What are you thinking for a fix? The most obvious one that jumps out to me is to have a box point to all the boxes for which it's the containing block (so it doesn't have to traverse the whole subtree every time to find them).

yeah, I think it's either having a linked list of all contained abspos boxes for each paintable. Or, better idea: diverging paintable tree from layout tree by and making contained abspos boxes to be direct children of containing block.

**Comment by ebanner** [comment]
> Or, better idea: diverging paintable tree from layout tree by and making contained abspos boxes to be direct children of containing block.

That sounds like a cool idea.

The other case where the containing block comes into play that I came across (in addition to abspos boxes) is tables. All table elements have the table container as its containing block (so a TD won't cause overflow on a TR for example).

Do you imagine it working the same way for tables? That is, the table container would have TBODY, TR, TD etc. as its children in the paintable tree?
